frontends/p4/inlining.h

```diff
@@ -100,10 +100,52 @@ struct PerInstanceSubstitutions {
 struct InlineSummary : public IHasDbPrint {
     /// Various substitutions that must be applied for each instance
     struct PerCaller {  // information for each caller
+        /// 3-tuple identifying all the invocations of the subparser which can use the same
+        /// inlined states.
+        /// Consists of:
+        /// - Pointer to the declaration of the invoked subparser
+        /// - Pointer to the expression representing the transition statement from the state
+        ///   which invokes the subparser
+        /// - Pointer to a newly constructed IndexedVector of all parser statements
+        ///   which follow the subparser invocation (not including the transition statement)
+        typedef std::tuple<const IR::Declaration_Instance *,
+                           const IR::Expression *,
+                           const IR::IndexedVector<IR::StatOrDecl> *> InlinedInvocationInfo;
+
+        /// Hash for InlinedInvocationInfo used as a key for unordered_map
+        struct key_hash {
+            std::size_t operator() (const InlinedInvocationInfo &k) const {
+                std::ostringstream oss;
+                std::get<1>(k)->dbprint(oss);
+                for (auto e : *std::get<2>(k)) {
+                    e->dbprint(oss);
+                }
+                return std::hash<const IR::Declaration_Instance *>{}(std::get<0>(k)) ^
+                        std::hash<std::string>{}(oss.str());
+            }
+        };
+
+        /// Binary equality predicate for InlinedInvocationInfo used as a key for unordered_map
+        struct key_equal {
+            bool operator() (const InlinedInvocationInfo &v0,
+                    const InlinedInvocationInfo &v1) const {
+                return std::get<0>(v0) == std::get<0>(v1) &&
+                        std::get<1>(v0)->equiv(*std::get<1>(v1)) &&
+                        std::get<2>(v0)->equiv(*std::get<2>(v1));
+            }
+        };
+
         /// For each instance (key) the container that is intantiated.
         std::map<const IR::Declaration_Instance*, const IR::IContainer*> declToCallee;
         /// For each instance (key) we must apply a bunch of substitutions
         std::map<const IR::Declaration_Instance*, PerInstanceSubstitutions*> substitutions;
+        /// For each distinct invocation of the subparser identified by InlinedInvocationInfo
+        /// we store the name of the next caller parser state which is a new state replacing
+        /// the start state of the callee parser (subparser).
+        /// Transition to this state is used in case there is another subparser invocation
+        /// which has the equivalent InlinedInvocationInfo.
```

Comment:
What is "equivalent"? Can you define it precisely?

Reply:
"Equivalent" here has following meaning:
Two subparser invocations are equivalent when following conditions are met:
* subparser instance declarations are same for both invocations
* the sequences of parser statements following the subparser invocation are same for both invocations
* transition statement is same for both invocations

I will add the information what equality means here to the comment, once we agree on the other details, as it is possible that it will change here as well.